File: graph_go/graph.go
package main

// var emptyStruct struct{}

type Graph struct {
	edges         map[interface{}]map[interface{}]int
	inNeighbors   map[interface{}]map[interface{}]bool
	expectedNodes int
}

func NewGraph(expectedNodes int) *Graph {
	return &Graph{
		edges:         make(map[interface{}]map[interface{}]int, expectedNodes),
		inNeighbors:   make(map[interface{}]map[interface{}]bool, expectedNodes),
		expectedNodes: expectedNodes,
	}
}

func (graph *Graph) Connect(vertexFrom interface{}, vertexTo interface{}, weight int) {
	if _, ok := graph.edges[vertexFrom]; !ok {
		graph.edges[vertexFrom] = make(map[interface{}]int, graph.expectedNodes)
		graph.inNeighbors[vertexFrom] = make(map[interface{}]bool, graph.expectedNodes)
	}

	if _, ok := graph.edges[vertexTo]; !ok {
		graph.edges[vertexTo] = make(map[interface{}]int, graph.expectedNodes)
		graph.inNeighbors[vertexTo] = make(map[interface{}]bool, graph.expectedNodes)
	}

	graph.edges[vertexFrom][vertexTo] = weight
	graph.inNeighbors[vertexTo][vertexFrom] = true
}

func (graph *Graph) FindAllOutNeighbors(vertex interface{}) map[interface{}]bool {
	keys := make(map[interface{}]bool, len(graph.edges))

	for key := range graph.edges[vertex] {
		keys[key] = true
	}

	return keys
}

func (graph *Graph) FindAllInNeighbors(vertex interface{}) map[interface{}]bool {
	keys := make(map[interface{}]bool, len(graph.edges))

	for key := range graph.inNeighbors[vertex] {
		keys[key] = true
	}

	return keys
}

func (graph *Graph) FindAllVertices() map[interface{}]bool {
	keys := make(map[interface{}]bool, len(graph.edges))

	for key := range graph.edges {
		keys[key] = true
	}

	return keys
}

func (graph *Graph) visit(vertex interface{}, unvisited map[interface{}]bool, L *[]interface{}) {
	if _, ok := unvisited[vertex]; ok {
		delete(unvisited, vertex)

		for neighbor := range graph.FindAllOutNeighbors(vertex) {
			graph.visit(neighbor, unvisited, L)
		}

		*L = append(*L, vertex)
	}
}

func (graph *Graph) assign(vertex interface{}, root interface{}, components map[interface{}]map[interface{}]bool) {
	hasBeenAssigned := false

	for root := range components {
		if _, ok := components[root][vertex]; ok {
			hasBeenAssigned = true
			break
		}
	}

	if !hasBeenAssigned {
		if _, ok := components[root]; !ok {
			components[root] = make(map[interface{}]bool)
		}

		components[root][vertex] = true

		for neighbor := range graph.FindAllInNeighbors(vertex) {
			graph.assign(neighbor, root, components)
		}
	}
}

func (graph *Graph) FindConnectedComponents() map[interface{}]map[interface{}]bool {
	unvisited := graph.FindAllVertices()

	L := make([]interface{}, 0, len(unvisited))

	for vertex := range graph.FindAllVertices() {
		graph.visit(vertex, unvisited, &L)
	}

	components := make(map[interface{}]map[interface{}]bool)

	for i := len(L) - 1; i >= 0; i -= 1 {
		graph.assign(L[i], L[i], components)
	}

	return components
}

File: GraphCsharp/Domain/Graph.cs
using System.Collections.Generic;

namespace GraphCsharp.Domain
{
    public class Graph<T>
    {
        private readonly Dictionary<T, Dictionary<T, int>> Edges;
        private readonly Dictionary<T, ISet<T>> InNeighbors;
        private readonly int ExpectedNodes;

        public Graph(int expectedNodes)
        {
            ExpectedNodes = expectedNodes;
            Edges = new Dictionary<T, Dictionary<T, int>>(ExpectedNodes);
            InNeighbors = new Dictionary<T, ISet<T>>(ExpectedNodes);
        }

        public void Connect(T vertexFrom, T vertexTo, int weight)
        {
            if (!Edges.ContainsKey(vertexFrom))
            {
                Edges[vertexFrom] = new Dictionary<T, int>(ExpectedNodes);
                InNeighbors[vertexFrom] = new HashSet<T>(ExpectedNodes);
            }

            if (!Edges.ContainsKey(vertexTo))
            {
                Edges[vertexTo] = new Dictionary<T, int>(ExpectedNodes);
                InNeighbors[vertexTo] = new HashSet<T>(ExpectedNodes);
            }

            Edges[vertexFrom][vertexTo] = weight;
            InNeighbors[vertexTo].Add(vertexFrom);
        }

        public ISet<T> FindAllOutNeighbors(T vertex)
        {
            return new HashSet<T>(Edges[vertex].Keys);
        }

        public ISet<T> FindAllInNeighbors(T vertex)
        {
            return new HashSet<T>(InNeighbors[vertex]);
        }

        public ISet<T> FindAllVertices()
        {
            return new HashSet<T>(Edges.Keys);
        }

        public Dictionary<T, ISet<T>> FindConnectedComponents()
        {
            // For each vertex u of the graph, mark u as unvisited
            var unvisited = FindAllVertices();

            // Let L be empty
            var L = new LinkedList<T>();

            // For each vertex u of the graph do Visit(u)
            foreach (var vertex in FindAllVertices())
            {
                Visit(vertex, unvisited, L);
            }

            var components = new Dictionary<T, ISet<T>>();

            // For each element u of L in order, do Assign(u,u)
            foreach (var vertex in L)
            {
                Assign(vertex, vertex, components);
            }

            return components;
        }

        private void Visit(T vertex, ISet<T> unvisited, LinkedList<T> L)
        {
            // If u is unvisited then
            if (unvisited.Contains(vertex))
            {
                // Mark u as visited
                unvisited.Remove(vertex);

                // For each out-neighbour v of u, do Visit(v)
                foreach (var neighbor in FindAllOutNeighbors(vertex))
                {
                    Visit(neighbor, unvisited, L);
                }

                // Prepend u to L
                L.AddFirst(vertex);
            }
        }

        private void Assign(T vertex, T root, Dictionary<T, ISet<T>> components)
        {
            var hasBeenAssigned = false;

            foreach (var component in components.Values)
            {
                if (component.Contains(vertex))
                {
                    hasBeenAssigned = true;
                    break;
                }
            }

            // If u has not been assigned to a component then
            if (!hasBeenAssigned)
            {
                if (!components.ContainsKey(root))
                {
                    components[root] = new HashSet<T>();
                }

                // Assign u as belonging to the component whose root is root
                components[root].Add(vertex);

                // For each in-neighbour v of u, do Assign(v,root)
                foreach (var neighbor in FindAllInNeighbors(vertex))
                {
                    Assign(neighbor, root, components);
                }
            }
        }
    }
}
